libraries/ui-library/src/utils/popup.ts

```diff
@@ -0,0 +1,124 @@
+      setPopupAsFixPosition(popup);
+      popup.style.top = '0px';
+    } else if (
+      window.innerHeight < popupDimensions.y + popupDimensions.height &&
+      window.innerHeight > popupDimensions.height &&
+      Math.abs(popupDimensions.y - window.innerHeight) <= popupDimensions.height
+    ) {
+      // handle case where popup is cut-off at the bottom but there is still space available above
+      // apply screen position to fixed popup
+      setPopupAsFixPosition(popup);
+      popup.style.top = `${window.innerHeight - popupDimensions.height}px`;
+    }
+
+    if (
+      window.innerWidth < popupDimensions.x + popupDimensions.width &&
+      window.innerWidth > popupDimensions.width &&
+      Math.abs(popupDimensions.x - window.innerWidth) <= popupDimensions.width
+    ) {
+      // handle case where popup is cut-off to the right
+      setPopupAsFixPosition(popup);
+      popup.style.left = `${window.innerWidth - popupDimensions.width}px`;
+    }
+  }, 0);
+}
+
+export function calcIsDropDownContentUp(
+  inputElement: HTMLElement,
+  wrapper: HTMLElement,
+  minPopupHeight: number
+): boolean {
+  const inputBoundingRect = inputElement.getBoundingClientRect();
+  const wrapperBoundingRect = wrapper.getBoundingClientRect();
+  const popupHeight = Math.max(wrapperBoundingRect.height, minPopupHeight);
+  const moreSpaceInTop = inputBoundingRect.y > window.innerHeight / 2;
+  return moreSpaceInTop && window.innerHeight < inputBoundingRect.bottom + popupHeight;
+}
+
+export function setPopupAsFixPosition(popup: HTMLElement): DOMRect {
+  const popupDimensions = popup.getBoundingClientRect();
+  popup.style.position = 'fixed';
+  popup.style.top = popupDimensions['top'] + 'px';
+  popup.style.left = popupDimensions['left'] + 'px';
+  popup.style.width = popupDimensions['width'] + 'px';
+  popup.style.height = popupDimensions['height'] + 'px';
+  return popupDimensions;
+}
+
+/*
+ * The position of the hoisted datepicker needs to be correctly calculated since the position changes to fixed.
+ * Thus if the user scrolls or adjusts the screen size we need to recalculate the datepicker position.
+ */
+export function movePopup(
+  hoist: boolean,
+  open: boolean,
+  popup: HTMLElement | undefined,
+  inputElement: HTMLElement | undefined,
+  wrapper: HTMLElement | undefined,
+  minPopupHeight: number
+) {
+  if (!hoist || !open || popup == null || inputElement == null || wrapper == null) {
+    return;
+  }
+  const popupBoundingClientRect = popup.getBoundingClientRect();
+  const popupHeight = popupBoundingClientRect.height;
+  const inputBoundingClientRect = inputElement.getBoundingClientRect();
+  const inputTop = inputBoundingClientRect.top;
+  const inputHeight = inputBoundingClientRect.height;
+
+  const isUp = calcIsDropDownContentUp(inputElement, wrapper, minPopupHeight);
+
+  if (isUp) {
+    popup.style.top = `${inputTop - popupHeight}px`;
+  } else {
+    popup.style.top = `${inputTop + inputHeight}px`;
+  }
```

Comment:
I simply removed the duplication between date- and timepicker, moving the common code into this file, and didn't have an in depth look at the logic.
I'd rather keep it as is, also because I don't see a significant difference in readability between the three options. Actually I even prefer the `if` version from a readability perspective.